src/messaging.cc

```diff
@@ -167,6 +176,43 @@ bool MessageEncoder::encode_array(Array* object, int size) {
   return true;
 }
 
+bool MessageEncoder::encode_map(Instance* instance) {
+  write_uint8(TAG_MAP);
+  Object* bk = instance->at(Instance::MAP_BACKING_OFFSET);
+  if (is_smi(bk)) return false;
+  HeapObject* backing = HeapObject::cast(bk);
+  Object* sz = instance->at(Instance::MAP_SIZE_OFFSET);
+  if (!is_smi(sz)) return false;
+  word size = Smi::cast(sz)->value();
+  write_cardinal(size);
+  if (size == 0) return true;  // Do this before looking at the backing, which may be null.
+  Smi* class_id = backing->class_id();
+  if (class_id == _program->list_class_id()) {
+    bk = Instance::cast(backing)->at(Instance::LIST_ARRAY_OFFSET);
+    if (is_smi(bk)) return false;
+    backing = HeapObject::cast(bk);
+  }
+  class_id = backing->class_id();
+  if (class_id != _program->array_class_id()) {
+    if (class_id == _program->large_array_class_id()) {
+      printf("Can't serialize large map\n");
+    }
+    return false;
+  }
+  Array* array = Array::cast(backing);
+  int i = 0;
```

Comment:
What are you using this for? You could ASSERT that it is the same as size after the loop?